src/ota_handler.rs

```diff
@@ -0,0 +1,352 @@
+        request_uuid: &str,
+    ) -> Result<(), DeviceManagerError> {
+        info!("Got update event");
+
+        self.send_ota_response(sdk, request_uuid, OTAStatus::InProgress)
+            .await?;
+
+        let path = std::path::Path::new(&self.download_file_path).join("update.bin");
+        let path = path.to_str().ok_or_else(|| {
+            DeviceManagerError::FatalError("wrong download file path".to_string())
+        })?;
+
+        wget(request_url, path).await?;
+
+        let bundle_info = self.rauc.info(path).await?;
+        debug!("bundle info: {:?}", bundle_info);
+
+        let compatible = self.rauc.compatible().await?;
+
+        if bundle_info.compatible != compatible {
+            error!(
+                "bundle '{}' is not compatible with system '{}'",
+                bundle_info.compatible, compatible
+            );
+            return Err(DeviceManagerError::UpdateError(
+                "bundle is not compatible".to_owned(),
+            ));
+        }
+
+        self.write_state_json(PersistentState {
+            uuid: request_uuid.to_string(),
+            slot: self.rauc.boot_slot().await?,
+        })?;
+
+        self.rauc
+            .install_bundle(path, std::collections::HashMap::new())
+            .await?;
+
+        debug!(
+            "install_bundle done, last_error={:?}",
+            self.last_error().await
+        );
+
+        debug!("rauc operation = {}", self.rauc.operation().await?);
+
+        let mut completed_update = self.rauc.receive_completed().await?;
+
+        info!("Waiting for signal...");
+        if let Some(completed) = completed_update.next().await {
+            let signal = completed.args().unwrap();
+            let signal = *signal.result();
+
+            info!("Completed signal! {:?}", signal);
+
+            match signal {
+                0 => {
+                    info!("Update successful");
+                    OTAHandler::reboot().await?;
+                }
+                _ => {
+                    error!("Update failed with signal {signal}");
+                    return Err(OTAError::Deploy.into());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn reboot() -> Result<(), DeviceManagerError> {
+        info!("Rebooting in 5 seconds");
+
+        if std::env::var("DM_DRYRUN").is_ok() {
+            info!("Dry run, exiting");
+
+            std::process::exit(0);
+        }
+
+        tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+
+        // TODO: use systemd api
+        let mut cmd = std::process::Command::new("shutdown");
+        cmd.arg("-r");
+        cmd.arg("now");
+
+        let output = cmd.output()?;
+
+        if output.status.success() && output.stderr.is_empty() {
+            panic!("Reboot command was successful, bye");
+        } else {
+            error!("Reboot failed {:?}", output.stderr);
+        }
+
+        Ok(())
+    }
+
+    pub async fn check_pending_ota_response(
```

Comment:
`check_pending_ota_response` is a bit unclear on side effects. Maybe `ensure_pending_ota_response()?